Security  구현 완료

SigLee2247 opened the pull request:
## What is this PR

- Security 구현 및 회원 가입 서비스 구현
## Changes

- AuthenticationFilter 에서 Redis 저장 시 더 좋은 방식이 있는지

## Issue Number

feat #1  

resolve #

close #

SigLee2247 committed:
feat: passwordEncoder Bean 등록 #1
암호화를 위한 passwordEncoder 빈 등록

SigLee2247 committed:
feat: 회원 저장을 위한 UserMapper 및 DTO 구현 #1
- toEntity(), toResponseIdDto()메서드 구현
- saveUser()를 위한 파라미터 및 반환 값 DTO 구현

SigLee2247 committed:
feat: 회원 저장을 위한 Service Repository 메서드 구현 #1
- 중복 username 확인을 위한 validUsernameExist() 구현
- 중복 사용자 조회를 위해 UserRepository findUserByUserName()메서드 구현

SigLee2247 committed:
feat: 사용자 테스트를 위한 Config 설정 #1
- UserService 테스트를 위한 Config 구현

SigLee2247 committed:
feat: 회원 저장 Service 단위 테스트 #1
- 단위 테스트를 위한 스텁데이터 구현 UserMock
- 회원 저장 성공 실패 단위 테스트 구현

SigLee2247 committed:
[fit]: 회원 저장 response mapper 위치 변경 #1
- response Wrapping 위치 service 변경 및 이에 따른 테스트 코드 변경

SigLee2247 committed:
feat: CommonPassword valid 어노테이션 구현 #1
- 자주 사용되는 password 검증을 위한 어노테이션 구현
- ConcurrentHashMap을 사용해 동시성 이슈 방지

SigLee2247 committed:
feat: csrf 에러 처리를 위한 csrf 설정 추가 #1

SigLee2247 committed:
feat: 비밀번호 와 다른 정보 같은지 검증하는 어노테이션 #1
- 비밀번호와 다른 정보들을 비교해 문제가 있는지 확인하기 위해 구현한 Custom Annotation

SigLee2247 committed:
feat: UserPostRequestDto Valid 적용 #1
- 전체 길이 10 자리 이상 구현
- 문자 숫자 특수문자 중 2가지 종류 이상 사용
- 한글 사용 금지
- 3회 이상연속된 문자 사용 불가능
- 자주 사용되는 비밀번호 사용 불가
- 비밀번호와 다른 정보들이 같은 정보인지 검증

SigLee2247 committed:
feat: 회원 등록 Controller 구현 #1
- Uri 구현을 위한 UriCreator 구현

SigLee2247 committed:
move: validation 이동 #1
- validation 폴더 생성 및 하위 폴더로 validation 클래스 이동
     - CommonPasswords
     - CommonPasswordValid
     - CommonPasswordValidator
     - PasswordInfoSameValid
     - PasswordInfoSameValidator

SigLee2247 committed:
chore: jjwt 의존성 버전 수정 #1
- 0.11.5 ->0.12.13 버전 수정

SigLee2247 committed:
feat: ymlProperties 읽어 오는 class 구현 #1
- @PropertySource 활용
    - YamlPropertySourceFactory 구현
    - yml 데이터 추가

SigLee2247 committed:
feat: Principal을 위한 객체 구현 및 UserDetailsService 구현 #1
- Principal 역할을 하는 Principal 구현
- UserDetailsServiceImpl 구현

SigLee2247 committed:
feat: JwtProvider 구현 및 테스트 코드 작성 #1
- claims에 List 정보가 아닌 String 정보 전달
- accessToken, refreshToken 구현 테스트

SigLee2247 committed:
feat: 레디스 설정 및 Test 작성 #1
- RedisTemplate 을 활용한 Redis 설정

SigLee2247 committed:
feat: 인증을 위한 AuthenticationFilter 구현 #1
- UsernamePasswordAuthenticationFilter 상속해 JSON 데이터 전달 받아 인증 진행
- login을 위한 LoginDto 구현

SigLee2247 committed:
feat: 시큐리티 예외 처리를 위한 ErrorResponse 구현 #1
- 예외 처리 클래스 미구현으로 인한 예외 처리 및 전달 Response 구현
- 예외 구현 시 제거

SigLee2247 committed:
feat: 인가를 위한 VerificationFilter 구현 #1
- 인가를 위한 VerificationFilter 구현
- OncePerRequestFilter 구현을 통한 Filter 구현

SigLee2247 committed:
fit: 데이터 properties 참조 #1
- redis 저장 시간 properties에서 가져옴
- accesstoken prefix properties에서 가져옴

SigLee2247 committed:
feat: 인증 실패 Handler 구현 #1
- 인증 실패 시 인증 실패 데이터 전달을 위한 Handler 구현

SigLee2247 committed:
feat: 인가 실패 EntryPoint 구현 #1
- 인가 실패 시 각 상태에 맞는 Response 구현

SigLee2247 committed:
feat: 로그아웃 성공 Handler 구현 #1
- 로그아웃 시 redis에서 refreshtoken 삭제 및
- cookie 안의 refresh 만료

SigLee2247 committed:
feat: 구현한 시큐리티 Config 추가 #1
- 인증 인가 Filter 추가
- 시큐리티 설정 추가

SigLee2247 committed:
fit: cookie 저장 시간 refreshToken 만료시간과 동일하게 수정 #1

SigLee2247 committed:
style: 불필요한 공백 제거 #1

SigLee2247 committed:
style: private 메서드 위치 변경 #1
- 클래스 최상단에는 public 메서드가 보이도록 위치 변경
- 대상 클래스
    - AuthenticationEntryPointHandler
    - LogoutSuccessCustomHandler

SigLee2247 committed:
feat: UserService 예외 처리를 위한 ExceptionAdvice 구현 #1
- 서비스 로직 에러 AOP
- Dto 검증 에러

SigLee2247 committed:
update: redis 저장 데이터 변경 #1
- 기존 : userName
- 변경 : Userinfo 객체의 String 값
- 변경 사유 : accessToken 재발급 시 필요한 권한이 User 엔티티에 존재하지 않기 떄문에 Redis에 저장 이후 재발급 시 사용하기 위해

SigLee2247 committed:
test: redis 데이터 중복 저장 테스트 #1
- 같은 key에 연속 저장 테스트

SigLee2247 committed:
feat: accessToken 갱신 서비스 구현 #1
- 같은 key에 연속 저장 테스트

SigLee2247 committed:
test: 로그인 테스트 #1
- 로그인 성공 테스트
- 로그인 실패 테스트
- 테스트를 위한 Config 설정
- passwordEncoder Bean등록 분리

SigLee2247 committed:
test: 토큰 갱싱 테스트 #1
- 로그인 성공 테스트
- 실패 테스트
- 테스트를 위한 데이터 추가

SigLee2247 committed:
test: 회원 가입 Controller 테스트 #1
- 회원 가입 성공 테스트
- 회원 가입 시 요청 DTO 에러 테스트

SigLee2247 committed:
test: Test Config 어노테이션 변경 #1
- 기존의 @Configuration 에서 @TestConfiguration 변경

jkde7721 commented on src/main/java/com/wanted/common/exception/GlobalExceptionAdvice.java:
```diff
@@ -0,0 +1,43 @@
+        log.error("BusinessLogic Exception Error : {}", e.getMessage());
+        final ErrorResponse errorResponse = ErrorResponse.of(e.getHttpStatus());
+        return new ResponseEntity(errorResponse, e.getHttpStatus());
+    }
```
```java
public ResponseEntity<ErrorResponse> handleBusinessException(CommonException e) {
    //...
    return new ResponseEntity<>(errorResponse, e.getHttpStatus());
}

이렇게 명시적으로 제네릭 지정해주시면 좋을 것 같습니다!
```

SigLee2247 replied:
제네릭 타입 명시적인 구현 완료했습니다!

jkde7721 commented on src/main/java/com/wanted/common/security/controller/AuthController.java:
```diff
@@ -0,0 +1,24 @@
+    private final AuthService service;
+
+    @PostMapping("/reissue")
+    public ResponseEntity reissue(HttpServletRequest request, HttpServletResponse response) {
```
여기도 마찬가지로 `ResponseEntity<Void>` 명시적으로 제네릭 지정하면 좋을 것 같습니다.

SigLee2247 replied:
제네릭 타입 명시적 구현 완료했습니다.

jkde7721 commented on src/main/java/com/wanted/common/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,92 @@
+        StringBuilder sb = new StringBuilder();
+        list.forEach(data -> sb.append(data).append(","));
+        return sb.deleteCharAt(sb.length() - 1).toString();
+    }
```
만약 `list`가 비어 있다면 `sb.deleteCharAt(sb.length() - 1)` 부분에서 StringIndexOutOfBoundsException 예외가 발생해 스프링이 유틸로 제공하는 
`StringUtils.collectionToCommaDelimitedString(list);` (내부적으로 StringBuilder 사용) 사용하는 것이 더 좋지 않을까 생각합니다.

SigLee2247 replied:
StringUtils에 대한 학습 진행 및 변경 완료 했습니다.
[공식 문서 링크](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/StringUtils.html)
감사합니다

jkde7721 commented on src/main/java/com/wanted/common/security/filter/JwtVerificationFilter.java:
```diff
@@ -0,0 +1,76 @@
+    private Principal createUserDetails(HttpServletRequest request) {
+        String token = getAuthenticationTokenToHeader(request).replace("Bearer ", "");
+        return new Principal(jwtProvider.getClaims(token));
+    }
```
`String` 객체에 대한 `replace` 메소드는 문자열의 시작~끝까지 "Bearer " 문자열을 찾고 이를 ""로 바꿔주는 방식으로 동작하는 것으로 알고 있는데, 
만약 (그럴 일은 아주 희박하겠지만) 토큰 자체에 "Bearer " 문자열이 포함되어 있다면 예상치 않게 동작할 수 있을 것 같습니다. 

따라서 `getAuthenticationTokenToHeader(request).substring(7);` 또는 

```java
String prefix = "Bearer ";
getAuthenticationTokenToHeader(request).substring(prefix.length());
```

이렇게 `substring` 메소드를 사용하는 것이 어떠신가요?

SigLee2247 replied:
해당 부분에 대한 변경 구현했습니다.
저 부분에 대한 생각을 하지 못했던 것 같습니다 수정했습니다!

jkde7721 commented on src/main/java/com/wanted/common/security/filter/JwtVerificationFilter.java:
```diff
@@ -0,0 +1,76 @@
+        filterChain.doFilter(request, response);
+    }
+
+    @Override
+    protected boolean shouldNotFilter(HttpServletRequest request) {
+        return !checkAuthenticationExist(request) || !checkAuthenticationBearerExist(request);
+    }
+
+    private String getAuthenticationTokenToHeader(HttpServletRequest request) {
+        return request.getHeader(HttpHeaders.AUTHORIZATION);
+    }
+
+    private boolean checkAuthenticationExist(HttpServletRequest request) {
+        return getAuthenticationTokenToHeader(request) != null;
+    }
+    private boolean checkAuthenticationBearerExist(HttpServletRequest request) {
+        return getAuthenticationTokenToHeader(request).startsWith("Bearer ");
+    }
```
메소드를 분리하는 이유는 재사용성을 위해서 인데 제 생각에는 `Authorization 헤더가 null이 아님`, `Authorization 헤더의 값이 "Bearer " 로 시작함` 각각의 조건을 다른 메소드에서 확인하는 경우가 거의 없을 거라고 생각합니다.
즉 현재 `checkAuthenticationExist`, `checkAuthenticationBearerExist` 메소드들은 `shouldNotFilter` 메소드에서만 같이 호출되고 있기 때문에 아래와 같이 `checkAuthenticationExist`, `checkAuthenticationBearerExist` 메소드 2개를 1개로 합치는 건 어떠신가요?

```java
@Override
protected boolean shouldNotFilter(HttpServletRequest request) {
    return !validateAuthorizationHeader(request);
}

private String getAuthorizationHeader(HttpServletRequest request) {
    return request.getHeader(HttpHeaders.AUTHORIZATION);
}

private boolean validateAuthorizationHeader(HttpServletRequest request) {
    String authorizationHeader = getAuthorizationHeader(request);
    return authorizationHeader != null && authorizationHeader.startsWith("Bearer ");
}
```

LEEGIHO94 replied:
해당 부분을 하나로 합쳐서 확인을 해봤는데 
```java
    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        return getAuthenticationTokenToHeader(request) == null || !getAuthenticationTokenToHeader(request).startsWith(jwtProperties.getPrefix());
    }
```
위와 같은 방식으로 코드를 구현할 경우 코드의 가독성이 떨어지지 않을까 하는 걱정이 됩니다.
추천해주신 내부 변수를 사용하는 방법을 사용하게 된다면 코드의 가독성이 좋아지기 때문에 크게 문제가 될 것 같지 않지만 내부 변수 사용을 지양할 경우에는 어떤 방식이 더 좋은지 잘 몰겠습니다..

SigLee2247 replied:
해당 부분 가독성 있게 코드 수정했습니다. 감사합니다.

jkde7721 commented:
현재 리프레시 토큰 관련 쿠키를 처리하는 메소드인 `createCookie`, `validCookieExist`, `searchCookieProperties` 가
`LogoutSuccessCustomHandler`, `AuthService`, `JwtAuthenticationFilter` 각각의 클래스에서 2번 이상 중복되는데 
아래와 같이 별도의 클래스인 CookieUtils로 분리하는 것은 어떠신가요??

```java
@RequiredArgsConstructor
@Component
public class CookieUtils {

    private final JwtProperties jwtProperties;

    public Cookie createRefreshCookie(String key) {
        Cookie cookie = new Cookie("Refresh", key);
        cookie.setDomain("localhost");
        cookie.setHttpOnly(true);
        cookie.setSecure(true);
        cookie.setPath("/api/auth/reissue");
        cookie.setMaxAge(60 * jwtProperties.getRefreshTokenValidityInSeconds());
        return cookie;
    }

    public Cookie[] validCookiesExist(HttpServletRequest request) {
        return Optional.ofNullable(request.getCookies()).orElse(new Cookie[]{});
    }

    public Optional<Cookie> searchCookieProperties(Cookie[] cookies, String name) {
        return Arrays.stream(cookies)
            .filter(cookie -> cookie.getName().equals(name))
            .findFirst();
    }

    public void deleteCookie(Cookie cookie) {
        cookie.setMaxAge(0);
    }
}
```

꼭 이렇게 분리하지는 않더라도 쿠키 처리 관련해서는 별도의 클래스로 분리한다면 
`LogoutSuccessCustomHandler`, `AuthService`, `JwtAuthenticationFilter` 클래스에서 쿠키 처리 관련 메소드들이 제거되어
각 클래스의 책임이 더 명확해지지 않을까 생각합니다!

jkde7721 commented:
현재 전체적으로 `InputStream` 또는 `String` → 객체, 객체 → `String`으로 바꾸는 작업이 많은데 
각 클래스 내에서 이 작업을 메소드로 분리해주기 보단, 아래와 같이 아예 별도의 Util 클래스로 분리하는 것도 좋지 않을까 생각합니다.
이렇게 되면 `ObjectMapperUtils` 클래스 내부에서 try-catch로 예외 처리를 해주기 때문에 `JwtAuthenticationFilter` 클래스의 `@SneakyThrows` 어노테이션을 사용할 필요도 없게 됩니다. 

다만 걱정되는 것은 제가 아직 전체 코드를 파악한 상태는 아니라서 이렇게 별도의 클래스로 분리하게 되면 어떤 식으로 영향이 미칠지 잘 모르겠어서, 아래 코드는 참고하시고 개인 판단으로 반영해주시면 될 것 같습니다. 

```java
@RequiredArgsConstructor
@Component
public class ObjectMapperUtils {

    private final ObjectMapper objectMapper;

    public <T> T transObject(InputStream src, Class<T> valueType) {
        try {
            return objectMapper.readValue(src, valueType);
        } catch (IOException e) {
            throw new CommonException(HttpStatus.INTERNAL_SERVER_ERROR, "JSON 변환 에러");
        }
    }

    public <T> T transObject(String content, Class<T> valueType) {
        try {
            return objectMapper.readValue(content, valueType);
        } catch (IOException e) {
            throw new CommonException(HttpStatus.INTERNAL_SERVER_ERROR, "JSON 변환 에러");
        }
    }

    public String transString(Object value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JsonProcessingException e) {
            throw new CommonException(HttpStatus.INTERNAL_SERVER_ERROR, "JSON 변환 에러");
        }
    }
}
```

jkde7721 commented on src/main/resources/application.yml:
```diff
@@ -22,3 +22,9 @@ logging:
     org.hibernate.type.descriptor.sql: trace
+
+jwt:
+  prefix: Bearer
```
jwt:
    prefix: 'Bearer '

이렇게 prefix에 공백을 포함시키면

`JwtVerificationFilter` 클래스에서 "Bearer " 리터럴로 사용하는 부분도 제거할 수 있을 것 같고
`JwtAuthenticationFilter` 클래스의 기존 `response.setHeader(HttpHeaders.AUTHORIZATION, jwtProperties.getPrefix() + " " + accessToken);` 코드도 `response.setHeader(HttpHeaders.AUTHORIZATION, jwtProperties.getPrefix() + accessToken);` 

이렇게 수정할 수 있지 않을까 합니다.

SigLee2247 replied:
구현 완료했습니다! 감사합니다!!

jkde7721 commented:
현재 모든 코드를 꼼꼼히 다 보지는 못했지만 전체적으로 봤을 때 깔끔하고 잘 작성된 코드라고 생각합니다. 
`PropertySourceFactory` 등과 같이 몰랐던 내용들을 코드를 통해 공부할 수 있어 좋았습니다.
다만 개인적으로 코드를 읽으며 조금 아쉬운 점이 있다면 메소드가 너무 분리되어 있어 코드 흐름을 한번에 파악하기 어려웠다는 점입니다. 예를 들어 

`AuthService` 클래스의 `reissue` 메소드 내의 `findUserInfo(findCookie(request))`는 현재 토큰 재발급을 요청한 사용자의 정보를 조회하는 기능을 합니다. `findUserInfo` 메소드의 내용을 파악하기 위해 다음과 같은 흐름으로 코드를 읽었습니다. 
 
findCookie → validateCookiesExist → searchCookieProperties → findUserInfo → transUserInfo → findAndDeleteToRedis 
→ findTokenToRedis → deleteToken 

한 기능의 구현을 파악하기 위해 스크롤을 옮겨 다녀야만 해서 제 느낌 상으로는 코드 문맥이 끊기는 느낌이 조금 있었습니다.
물론 메소드명이 잘 작성되어 있어 굳이 메소드의 내부 내용을 읽어보지 않아도 되지만, 이 경우에는 메소드 바디 양이 3~5줄 정도 밖에 되지 않아 몇몇 메소드들은 하나로 합쳐도 되지 않나 라는 생각이 들었습니다. (예를 들어 `AuthService`의 `deleteToken` 메소드)
물론 사람마다 의견이 갈릴 수 있는 부분이라고 생각합니다. 아무튼 짧은 시간동안 시큐리티 구현하시느라 수고 많으셨습니다!👍

SigLee2247 committed:
move: PasswordEncoderConfig 위치 변경 #1
- property -> config 위치 변경

SigLee2247 committed:
fit: List -> String 변환 메서드 변경 #1
- StringUtils.collectionToCommaDelimitedString(list) 활용으로 수정

SigLee2247 committed:
refactor: setAuthenticationToContext 메서드 인라인화 #1

SigLee2247 committed:
refactor: 직접 사용 하던 Bearer 문자 JwtProperties 참조 수정 #1

SigLee2247 committed:
refactor: shouldNotFilter 메서드 메서드 병합 #1
- 기존의 checkAuthenticationExist, checkAuthenticationBearerExist 로 쪼개져있는 코드 하나의 메서드로 병합

SigLee2247 committed:
test: jwtProperties 테스트 결과 수정 #1
yml 작성 데이터 수정으로 인한 데이터 수정

SigLee2247 committed:
style: AuthControllerTest 코드 스타일 수정 #1

SigLee2247 committed:
refactor: AuthService 토큰 생성 시 JwtProperties 에서 조회 #1

SigLee2247 committed:
refactor: 제네릭 타입 명시 #1

SigLee2247 committed:
refactor: Custom ObjectMapper 구현 #1
- 자주 사용하는 ObjectMapper의 예외 처리를 진행한 ObjectMapperUtils 구현
